0000-plugins.md

```diff
@@ -0,0 +1,169 @@
+
+# Reference-level explanation
+
+Plugins are loaded from `./plugins/*.wasm` at startup. The endpoint `/api/v4/site` gets a new field `active_plugins`.
+
+## Plugin Hooks
+
+Lemmy will have to add hooks for specific actions, which can then be used by plugins. In general there are two types of hooks: before an action is written to the database, so it can be rejected or altered. And after it is written to the database, mainly for different types of notifications.
+
+Data is passed to plugins using the existing structs linked below, serialized to JSON.
+
+For the initial implementation, the following hooks will be available:
+
+### Before writing to Database
+
+These hooks can be used to reject or alter user actions based on different criteria.
+
+- Post
+    - [create_local_post](https://github.com/LemmyNet/lemmy/blob/0.19.7/crates/db_schema/src/source/post.rs#L67)
+    - [update_local_post](https://github.com/LemmyNet/lemmy/blob/0.19.7/crates/db_schema/src/source/post.rs#L98)
+    - [receive_federated_post](https://github.com/LemmyNet/lemmy/blob/0.19.7/crates/db_schema/src/source/post.rs#L67)
+    - [post_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/post.rs#L171)
+- Comment
+    - [create_local_comment](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L59)
+    - [update_local_comment](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L84)
+    - [receive_federated_comment](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L59)
+    - [post_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L119)
+
+### After writing to Database
+
+These are mainly useful to generate notifications.
+
+- Post
+	- [new_post](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/post.rs#L19)
+	- [new_post_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/post.rs#L157)
+- Comment
+	- [new_comment](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L27)
+	- [new_comment_vote](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/source/comment.rs#L105)
+
+## Example
+
+Below is a simple Go plugin which uses the `create_local_post` hook to replace `Rust` in post body with `Go`, and reject posts which mention `Java`. 
+
+Also checkout the documentation for Extism's [Go Plugin Development Kit](https://github.com/extism/go-pdk).
+
+```golang
+package main
+
+import (
+	"github.com/extism/go-pdk"
+	"strings"
+)
+
+type CreatePost struct {
+	Name string `json:"name"`
+	Body *string `json:"body,omitempty"`
+	...
+  }
+
+//export create_local_post
+func my_plugin() int32 {
+	params := CreatePost{}
+	// Read post data
+	err := pdk.InputJSON(&params)
+	if err != nil {
+		pdk.SetError(err)
+		return 1
+	}
+	if strings.Contains(params.Body, "Java") {
+		// Throw error to reject post
+		pdk.SetError("We dont talk about Java)
```

Comment:
We may want to compile the example code so that it does not contain syntax errors.